compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -116,11 +115,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected: Ty<'tcx>,
         expected_ty_expr: Option<&'tcx hir::Expr<'tcx>>,
     ) -> Ty<'tcx> {
+        debug!("check_expr_coercible_to_type");
         let ty = self.check_expr_with_hint(expr, expected);
         // checks don't need two phase
-        self.demand_coerce(expr, ty, expected, expected_ty_expr, AllowTwoPhase::No)
+        let (res_ty, diag) =
+            self.demand_coerce_diag(expr, ty, expected, expected_ty_expr, AllowTwoPhase::No);
+
+        if let Some(diag) = diag {
+            self.note_caller_chooses_ty_for_ty_param(diag, expected, ty);
```

Comment:
I don't think this is the right place since `check_expr_coercible_to_type` does not only get run on return expressions but almost anywhere where coercion can happen. See my comment over at file `issue-67945-1.full.stderr`.

I think you should move the `err.subdiagnostic(NoteCallerChoosesTyForTyParam { … }) out of `try_suggest_return_impl_trait` and into the `hir::FnRetTy::Return` branch. I that makes sense.

Reply:
Moved the suggestion into the `hir::FnRetTy::Return` branch.